v2transport/transport.go

```diff
@@ -0,0 +1,625 @@
+		receivedPrefix = p.receivedPrefix
+	}
+
+	recvData, err := p.Receive(64 - len(receivedPrefix))
+	if err != nil {
+		return err
+	}
+
+	var ellswiftTheirs [64]byte
+
+	if initiating {
+		// If we are initiating, read all 64 bytes into ellswiftTheirs.
+		copy(ellswiftTheirs[:], recvData)
+	} else {
+		// If we are the responder, then we need to account for the bytes
+		// already received as part of matching against the starting v1
+		// transport bytes. We sanity check receivedPrefix in case it is too
+		// large for some reason.
+		prefixLen := len(receivedPrefix)
+		if prefixLen > 16 {
+			return errPrefixTooLarge
+		}
+
+		copy(ellswiftTheirs[:], receivedPrefix)
+		copy(ellswiftTheirs[prefixLen:], recvData)
+	}
+
+	// Calculate the v1 protocol's message prefix and see if the bytes read
+	// read into ellswiftTheirs matches it.
+	v1Prefix := createV1Prefix(net)
+
+	// ellswiftTheirs should be at least 16 bytes if receive succeeded, but
+	// just in case, check the size.
+	if len(ellswiftTheirs) < 16 {
+		return errInsufficientBytes
+	}
+
+	if !initiating && bytes.Equal(ellswiftTheirs[4:16], v1Prefix[4:16]) {
+		return errWrongNetV1Peer
+	}
+
+	// Calculate the shared secret to be used in creating the packet ciphers.
+	ecdhSecret, err := btcec.V2Ecdh(
+		p.privkeyOurs, ellswiftTheirs, p.ellswiftOurs, initiating,
+	)
+	if err != nil {
+		return err
+	}
+
+	err = p.createV2Ciphers(ecdhSecret[:], initiating, net)
+	if err != nil {
+		return err
+	}
+
+	// Send garbage terminator.
+	p.Send(p.sendGarbageTerm)
+
+	// Optionally send decoy packets after garbage terminator.
+	aad := p.sentGarbage
+	for i := 0; i < len(decoyContentLens); i++ {
+		decoyContent := make([]byte, decoyContentLens[i])
+		encPacket, _, err := p.V2EncPacket(decoyContent, aad, true)
+		if err != nil {
+			return err
+		}
+
+		p.Send(encPacket)
+
+		// TODO: Does this cause anything to leak?
+		aad = nil
+	}
+
+	// Send version packet.
+	_, _, err = p.V2EncPacket(transportVersion, aad, false)
+	if err != nil {
+		return err
+	}
+
+	// Skip garbage until encountering garbage terminator.
+	recvGarbage, err := p.Receive(16)
+	if err != nil {
+		return err
+	}
+
+	// The BIP text states that we can read up to 4111 bytes. We've already
+	// read 16 bytes for the garbage terminator, so we need to read up to
+	// maxGarbageLen more bytes.
+	for i := 0; i < maxGarbageLen; i++ {
+		recvGarbageLen := len(recvGarbage)
+		if bytes.Equal(recvGarbage[recvGarbageLen-16:], p.recvGarbageTerm) {
+			_, err = p.V2ReceivePacket(recvGarbage[:recvGarbageLen-16])
+			return err
+		}
+
+		recvData, err := p.Receive(1)
+		if err != nil {
+			return err
+		}
+		recvGarbage = append(recvGarbage, recvData...)
+	}
+
+	// Return an error if the garbage terminator was not received after 4 KiB.
+	return errGarbageTermNotRecv
+}
+
+// V2EncPacket takes the contents and aad and returns a ciphertext.
+func (p *Peer) V2EncPacket(contents []byte, aad []byte, ignore bool) ([]byte,
```

Comment:
Would be good to create some basic benchmarks for:
  * handshake creation
  * handshake complete 
  * v2 encrypt 
  * v2 decrypt 

Just so we can get an ideal re the current cpu/memory utilization.